ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -221,6 +190,64 @@ let clean_local_resources () : unit =
     error "%s : %s" msg (ExnHelper.string_of_exn e) ;
     raise (Auth_service_error (E_GENERIC, msg))
 
+module Winbind = struct
+  let name = "winbind"
+
+  let is_ad_enabled ~__context =
+    ( Helpers.get_localhost ~__context |> fun self ->
+      Db.Host.get_external_auth_type ~__context ~self )
+    |> fun x -> x = Xapi_globs.auth_type_AD
+
+  let start ~timeout ~wait_until_success =
+    Xapi_systemctl.start ~timeout ~wait_until_success name
+
+  let restart ~timeout ~wait_until_success =
+    Xapi_systemctl.restart ~timeout ~wait_until_success name
+
+  let stop ~timeout ~wait_until_success =
+    Xapi_systemctl.stop ~timeout ~wait_until_success name
+
+  let configure () =
+    (* Refresh winbind configuration to handle upgrade from PBIS
+     * The winbind configuration needs to be refreshed before start winbind daemon *)
+    let domain = get_service_name () in
+    let workgroup = query_domain_workgroup domain in
```

Comment:
Both of these functions create a context, which is slightly inefficient because they could actually share the same one

Reply:
will refine into one single function to query both from db.